File: content/problems/0094-binary-tree-inorder-traversal/solutions/iterative-stack/Solution.kt
class Solution {
    fun inorderTraversal(root: TreeNode?): List<Int> {
        val result = mutableListOf<Int>()
        val stack = java.util.ArrayDeque<TreeNode>()
        var current = root
        while (current != null || stack.isNotEmpty()) {
            while (current != null) {
                stack.push(current)
                current = current.left
            }
            current = stack.pop()
            result.add(current.`val`)
            current = current.right
        }
        return result
    }
}

File: content/problems/0094-binary-tree-inorder-traversal/solutions/iterative-stack/solution.cpp
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        stack<TreeNode*> nodes;
        TreeNode* current = root;
        while (current || !nodes.empty()) {
            while (current) {
                nodes.push(current);
                current = current->left;
            }
            current = nodes.top(); nodes.pop();
            result.push_back(current->val);
            current = current->right;
        }
        return result;
    }
};
